src/python/pants/core/target_types.py

```diff
@@ -42,19 +57,114 @@
+@frozen_after_init
+class HTTPSource:
+    url: str
+    len: int
+    sha256: str
+    # Defaults to last part of the URL path (E.g. `index.html`)
+    filename: str
+
+    def __init__(self, url: str, *, len: int, sha256: str, filename: str = ""):
+        for field in dataclasses.fields(self):
+            value = locals()[field.name]
+            if not isinstance(value, getattr(builtins, cast(str, field.type))):
+                raise TypeError(f"`{field.name}` must be a `{field.type}`, got `{type(value)!r}`.")
+
+        self.url = url
+        self.len = len
+        self.sha256 = sha256
+        self.filename = filename or self.url.rsplit("/", 1)[1]
+
+        if not self.filename:
+            raise ValueError(
+                "Couldn't deduce filename from `url`. Please specify the `filename` argument."
+            )
+        if "\\" in self.filename or "/" in self.filename:
+            raise ValueError("`filename` cannot contain a path separator.")
+
+
+class AssetSourceField(SingleSourceField):
+    value: str | HTTPSource  # type: ignore[assignment]
+
+    @classmethod
+    def compute_value(  # type: ignore[override]
+        cls, raw_value: Optional[Union[str, HTTPSource]], address: Address
+    ) -> Optional[Union[str, HTTPSource]]:
+        if raw_value is None or isinstance(raw_value, str):
+            return super().compute_value(raw_value, address)
+        elif not isinstance(raw_value, HTTPSource):
+            raise InvalidFieldTypeException(
+                address,
+                cls.alias,
+                raw_value,
+                expected_type="a string or an `http_source` object",
+            )
+        return raw_value
+
+    def validate_resolved_files(self, files: Sequence[str]) -> None:
+        if isinstance(self.value, str):
+            super().validate_resolved_files(files)
+
+    @property
+    def globs(self) -> tuple[str, ...]:
+        if isinstance(self.value, str):
+            return (self.value,)
+        return ()
+
+    @property
+    def file_path(self) -> str:
+        assert self.value
+        filename = self.value if isinstance(self.value, str) else self.value.filename
+        return os.path.join(self.address.spec_path, filename)
+
+
+@rule_helper
+async def _hydrate_asset_source(request: GenerateSourcesRequest) -> GeneratedSources:
+    target = request.protocol_target
+    source_field = target[AssetSourceField]
+    if isinstance(source_field.value, str):
+        return GeneratedSources(request.protocol_sources)
+
+    http_source = source_field.value
+    file_digest = FileDigest(http_source.sha256, http_source.len)
+    # NB: This just has to run, we don't actually need the result because we know the Digest's
+    # FileEntry metadata.
+    await Get(Digest, DownloadFile(http_source.url, file_digest))
+    snapshot = await Get(
+        Snapshot,
+        CreateDigest(
+            [
+                FileEntry(
+                    path=source_field.file_path,
```

Comment:
Do you know what will happen if the `filename` is wrong? Cryptic engine error? If so, maybe we should try catching that and enrich the error.

Reply:
What do you mean by wrong?

Reply:
Like the filename that gets downloaded is actually `foo.py` but I typed it as `f.py` on accident

Reply:
Oh I think you misunderstand filename. If it's specified it's the name we'll use. It's not just when it can't be defaulted from the URL path, but also for renaming.

So if you specify it, it can't ever be "wrong".

Reply:
Got it. Oh that's a good point, thoughts on editing `help` for the `source` field to mention how this all works?

Reply:
Oh yeah, gotta do that 😅